### PR TITLE
remove background token renew

### PR DESCRIPTION
**Pull request by hensur**
There are some edge cases where a background token renew wouldn't work
and the client is left without a valid token.
This usually happens when vault is restarted.

These cases (and the usual token renew) can be handled by catching
failed request with StatusCode 403 and retrying them, if a new
authentication attempt succeeds.

This way, all denied requests will be retried with a new valid
token. Therefore, there is no need to renew tokens in the background.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -53,55 +74,89 @@ func NewClient(addr string, tlsConf *TLSConfig, opts ...ClientOpts) (*Client, er
...
+	if err != nil {
+		return err
+	}
+
+	c.SetToken(res.Auth.ClientToken)
+
+	return nil
+}
+
+func (c *Client) Request(method string, path []string, body, response interface{}, opts *RequestOptions) error {
```
Comment (martin-helmich) [inline client.go:97]: Maybe add a `if opts == nil { opts = &RequestOptions{} }` to the top of the function? That way you can skip all the `if opts != nil` tests scattered throughout the function.

File: client.go [hunk 2]
```diff
@@ -53,55 +74,89 @@ func NewClient(addr string, tlsConf *TLSConfig, opts ...ClientOpts) (*Client, er
...
 	if err != nil {
-		return err
+		return errors.Wrap(err, "request failed")
 	}
-	defer resp.Body.Close()
 
-	if response != nil {
-		respBody, err := ioutil.ReadAll(resp.Body)
+	tokenRenewRequested := opts != nil && (opts.RenewToken == nil || *opts.RenewToken)
+	if resp.StatusCode == 403 && c.auth != nil && tokenRenewRequested {
```
Comment (martin-helmich) [inline client.go:117]:
```suggestion
	if resp.StatusCode == http.StatusForbidden && c.auth != nil && tokenRenewRequested {
```

File: client.go [hunk 3]
```diff
@@ -5,11 +5,32 @@ import (
...
+type RequestOptions struct {
+	Parameters url.Values
+
+	// RenewToken defines if the client should retry this Request with a new Token if it fails because of
+	// 403 Permission Denied
+	// The default behavior of the client is to always Request a new Token on 403
+	// Only if this is explicitly set to false, the client will continue processing the first failed request
+	// This should generally only be disabled for TokenAuth requests (a failed TokenAuth request can't be fixed by
+	// doing another TokenAuth request, this would lead to infinite recursion)
+	RenewToken *bool
```
Comment (martin-helmich) [inline client.go:33]: Might it be easier to turn this around and introduce a `SkipRenewal` parameter that has a null value of `false` and doesn't need to be a pointer?

Comment (hensur) [reply client.go]: This does indeed simplify the code a lot. I didn't want to introduce a double negative like `DisableRenewal`, but `SkipRenewal` sounds good.

File: client.go (hunk 1 above)
Comment (hensur) [reply client.go:99]: :+1:

**Review by martin-helmich — CHANGES_REQUESTED** [review]

File: client.go [hunk 4]
```diff
@@ -53,55 +76,92 @@ func NewClient(addr string, tlsConf *TLSConfig, opts ...ClientOpts) (*Client, er
...
 		}
 
-		if err = json.Unmarshal(respBody, response); err != nil {
-			return err
+		resp, err = c.RawRequest(r)
+		if err != nil {
+			return errors.Wrap(err, "request with new token failed")
 		}
 	}
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (hensur) [reply client.go:135]: The first statement in the if condition actually does this. The response is closed there and a new request with a new token is created.

[the remark this answers is not in the document]
Comment (martin-helmich) [reply client.go:135]: Jep. Nevermind, then.

**Review by martin-helmich — APPROVED** [review]

**Pull request merged**